File: WallpaperHub/UserProfile.swift
import Foundation
import AppKit

/// ユーザープロフィール
struct UserProfile: Codable, Equatable {
    let uid: String
    var displayName: String
    var email: String?
    var photoURL: String?
    var authProvider: AuthProvider
    var createdAt: Date
    var lastSyncAt: Date?

    /// 管理者フラグ
    var isAdmin: Bool = false

    /// カスタムアバター画像のローカルパス
    var customAvatarPath: String?

    /// 自己紹介文
    var bio: String?

    /// SNSリンク
    var socialLinks: SocialLinks

    /// 同期設定
    var syncPreferences: SyncPreferences

    enum AuthProvider: String, Codable {
        case anonymous
        case email
        case google
    }

    /// SNSリンク情報
    struct SocialLinks: Codable, Equatable {
        var twitter: String?
        var instagram: String?
        var pixiv: String?
        var skeb: String?
        var youtube: String?

        /// 設定されているリンクが1つ以上あるか
        var hasAnyLink: Bool {
            [twitter, instagram, pixiv, skeb, youtube].contains { $0?.isEmpty == false }
        }

        /// 各SNSリンクの必須プレフィックス
        private static let requiredPrefixes: [(KeyPath<SocialLinks, String?>, String)] = [
            (\.twitter, "https://x.com/"),
            (\.instagram, "https://www.instagram.com/"),
            (\.pixiv, "https://www.pixiv.net/users/"),
            (\.skeb, "https://skeb.jp/"),
            (\.youtube, "https://www.youtube.com/"),
        ]

        /// 入力されたリンクがすべて正しいURLプレフィックスを持っているか
        var isAllLinksValid: Bool {
            for (keyPath, prefix) in Self.requiredPrefixes {
                if let value = self[keyPath: keyPath], !value.isEmpty {
                    if !value.hasPrefix(prefix) {
                        return false
                    }
                }
            }
            return true
        }

        /// 指定フィールドのバリデーションエラーメッセージ（空欄ならnil）
        static func validationError(for value: String, prefix: String) -> String? {
            guard !value.isEmpty else { return nil }
            if !value.hasPrefix(prefix) {
                return "\(prefix) で始まるURLを入力してください"
            }
            return nil
        }
    }

    struct SyncPreferences: Codable, Equatable {
        var syncCollections: Bool = true
        var syncSchedules: Bool = true
        var syncEnvironmentRules: Bool = true
        var syncSettings: Bool = true
    }

    enum CodingKeys: String, CodingKey {
        case uid, displayName, email, photoURL, authProvider
        case createdAt, lastSyncAt, isAdmin
        case customAvatarPath, bio, socialLinks
        case syncPreferences
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        uid = try container.decode(String.self, forKey: .uid)
        displayName = try container.decode(String.self, forKey: .displayName)
        email = try container.decodeIfPresent(String.self, forKey: .email)
        photoURL = try container.decodeIfPresent(String.self, forKey: .photoURL)
        authProvider = try container.decode(AuthProvider.self, forKey: .authProvider)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        lastSyncAt = try container.decodeIfPresent(Date.self, forKey: .lastSyncAt)
        isAdmin = try container.decodeIfPresent(Bool.self, forKey: .isAdmin) ?? false
        customAvatarPath = try container.decodeIfPresent(String.self, forKey: .customAvatarPath)
        bio = try container.decodeIfPresent(String.self, forKey: .bio)
        socialLinks = try container.decodeIfPresent(SocialLinks.self, forKey: .socialLinks) ?? SocialLinks()
        syncPreferences = try container.decodeIfPresent(SyncPreferences.self, forKey: .syncPreferences) ?? SyncPreferences()
    }

    init(
        uid: String,
        displayName: String,
        email: String? = nil,
        photoURL: String? = nil,
        authProvider: AuthProvider,
        createdAt: Date,
        lastSyncAt: Date? = nil,
        isAdmin: Bool = false,
        customAvatarPath: String? = nil,
        bio: String? = nil,
        socialLinks: SocialLinks = SocialLinks(),
        syncPreferences: SyncPreferences = SyncPreferences()
    ) {
        self.uid = uid
        self.displayName = displayName
        self.email = email
        self.photoURL = photoURL
        self.authProvider = authProvider
        self.createdAt = createdAt
        self.lastSyncAt = lastSyncAt
        self.isAdmin = isAdmin
        self.customAvatarPath = customAvatarPath
        self.bio = bio
        self.socialLinks = socialLinks
        self.syncPreferences = syncPreferences
    }

    /// デフォルト初期化
    static func create(
        uid: String,
        displayName: String = "ユーザー",
        email: String? = nil,
        photoURL: String? = nil,
        authProvider: AuthProvider = .anonymous,
        isAdmin: Bool = false
    ) -> UserProfile {
        UserProfile(
            uid: uid,
            displayName: displayName,
            email: email,
            photoURL: photoURL,
            authProvider: authProvider,
            createdAt: Date(),
            lastSyncAt: nil,
            isAdmin: isAdmin,
            syncPreferences: SyncPreferences()
        )
    }

    // MARK: - アバター画像の取得

    /// カスタムアバター → Google/メールのphotoURL → nil の優先順位で画像を返す
    var resolvedAvatarImage: NSImage? {
        // カスタムアバターが設定されている場合はそちらを優先
        if let path = customAvatarPath, !path.isEmpty {
            // アバターパスの安全性を検証（ディレクトリ外のファイルアクセスを防止）
            let avatarDir = UserProfile.avatarDirectory.path
            guard path.hasPrefix(avatarDir) else { return nil }
            return NSImage(contentsOfFile: path)
        }
        return nil
    }

    /// アバター画像の保存ディレクトリ
    static var avatarDirectory: URL {
        guard let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else {
            let fallback = FileManager.default.temporaryDirectory.appendingPathComponent("WallBlank/Avatars", isDirectory: true)
            try? FileManager.default.createDirectory(at: fallback, withIntermediateDirectories: true)
            return fallback
        }
        let dir = appSupport.appendingPathComponent("WallBlank/Avatars", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }
}

File: rust-core/artia-core/src/models/user_profile.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// 認証プロバイダー
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AuthProvider {
    Anonymous,
    Email,
    Google,
}

/// SNSリンク情報
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SocialLinks {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub twitter: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub instagram: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pixiv: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub youtube: Option<String>,
}

impl SocialLinks {
    /// 設定されているリンクが1つ以上あるか
    pub fn has_any_link(&self) -> bool {
        [&self.twitter, &self.instagram, &self.pixiv, &self.youtube]
            .iter()
            .any(|link| link.as_ref().is_some_and(|s| !s.is_empty()))
    }

    /// 全リンクのバリデーション
    pub fn is_all_links_valid(&self) -> bool {
        let checks = [
            (&self.twitter, "https://x.com/"),
            (&self.instagram, "https://www.instagram.com/"),
            (&self.pixiv, "https://www.pixiv.net/users/"),
            (&self.youtube, "https://www.youtube.com/"),
        ];

        checks.iter().all(|(link, prefix)| {
            match link {
                Some(value) if !value.is_empty() => value.starts_with(prefix),
                _ => true,
            }
        })
    }
}

/// 同期設定
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncPreferences {
    #[serde(default = "default_true")]
    pub sync_collections: bool,
    #[serde(default = "default_true")]
    pub sync_schedules: bool,
    #[serde(default = "default_true")]
    pub sync_environment_rules: bool,
    #[serde(default = "default_true")]
    pub sync_settings: bool,
}

fn default_true() -> bool {
    true
}

impl Default for SyncPreferences {
    fn default() -> Self {
        Self {
            sync_collections: true,
            sync_schedules: true,
            sync_environment_rules: true,
            sync_settings: true,
        }
    }
}

/// ユーザープロフィール
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UserProfile {
    pub uid: String,
    pub display_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(rename = "photoURL", skip_serializing_if = "Option::is_none")]
    pub photo_url: Option<String>,
    pub auth_provider: AuthProvider,
    pub created_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_sync_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub is_admin: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_avatar_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bio: Option<String>,
    #[serde(default)]
    pub social_links: SocialLinks,
    #[serde(default)]
    pub sync_preferences: SyncPreferences,
}

impl UserProfile {
    /// デフォルト初期化
    pub fn create(uid: String, display_name: Option<String>, auth_provider: AuthProvider) -> Self {
        Self {
            uid,
            display_name: display_name.unwrap_or_else(|| "ユーザー".to_string()),
            email: None,
            photo_url: None,
            auth_provider,
            created_at: Utc::now(),
            last_sync_at: None,
            is_admin: false,
            custom_avatar_path: None,
            bio: None,
            social_links: SocialLinks::default(),
            sync_preferences: SyncPreferences::default(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_social_links_validation() {
        let valid = SocialLinks {
            twitter: Some("https://x.com/user".to_string()),
            instagram: None,
            pixiv: Some("https://www.pixiv.net/users/12345".to_string()),
            youtube: None,
        };
        assert!(valid.is_all_links_valid());
        assert!(valid.has_any_link());

        let invalid = SocialLinks {
            twitter: Some("https://twitter.com/user".to_string()),
            ..Default::default()
        };
        assert!(!invalid.is_all_links_valid());
    }

    #[test]
    fn test_user_profile_json_roundtrip() {
        let profile = UserProfile::create(
            "uid123".to_string(),
            Some("テストユーザー".to_string()),
            AuthProvider::Google,
        );
        let json = serde_json::to_string(&profile).unwrap();
        let decoded: UserProfile = serde_json::from_str(&json).unwrap();
        assert_eq!(decoded.uid, "uid123");
        assert_eq!(decoded.display_name, "テストユーザー");
    }
}
